### PR TITLE
Added loop option and loop size into configs

### PR DESCRIPTION
**Pull request by garvcho**

### CODE REVIEW

**Comment by ccelio** [comment]
Can you describe exactly what you're doing here? It looks like you're dedicating register x31 to be the loop count variable? And what are you looping over? The entire program?

**Comment by garvcho** [comment]
Yes, the intention was to add the option of having just one section instead of multiple sections and running it multiple times to avoid the overhead of jumping to different addresses. The header and foot aren't being looped though. The goal of this was to get more accurate power measurements of the main code without it being influenced by the overhead of setting everything else up.

File: config/default.config [hunk 1]
```diff
@@ -5,6 +5,8 @@ torture.generator.amo       true
 torture.generator.mul       true
 torture.generator.divider   true
 torture.generator.run_twice true
```
Comment (colinschmidt) [inline config/default.config]: Isn't this run_twice config obsolete now (as I recall it is basically a hardcoded loop_size=2)

**Comment by colinschmidt** [comment]
Hmm, I was hoping this would be a bit more general. For example, I was hoping you would be able to generate sequences of loops just like I can generate a sequence of ALU of FPU ops.

File: generator/src/main/scala/HWRegPool.scala [hunk 2]
```diff
@@ -84,12 +85,14 @@ trait PoolsMaster extends HWRegPool
...
 
-class XRegsPool extends ScalarRegPool
+class XRegsPool(loop: Boolean)  extends ScalarRegPool
 {
   val (name, regname, ldinst, stinst) = ("xreg", "reg_x", "ld", "sd")
   
   hwregs += new HWReg("x0", true, false)
-  for (i <- 1 to 31)
+  var last_reg = 31
+  if (loop) {last_reg = 30}
```
Comment (colinschmidt) [inline generator/src/main/scala/HWRegPool.scala]: I would prefer this to be done:

```
val last_reg = if (loop) 30 else 31
```

Then we don't need it to be a var as well.

**Comment by garvcho** [comment]
I could try to add that functionality as well but my immediate goal was to get an overall loop working because when I ran the generated assembly code on Primetime in the rocket chip repository, the power reading I got was noticeably smaller than I expected which is why I wanted to add the overall loop for reassurance that it wasn't lower because of the overhead of setting up the registers.

File: generator/src/main/scala/Prog.scala [hunk 3]
```diff
@@ -272,7 +272,7 @@ class Prog(memsize: Int, veccfg: Map[String,String], run_twice: Boolean)
       progsegs += ProgSeg()
 
     progsegs.last.insts += inst
-
+
```
Comment (colinschmidt) [inline generator/src/main/scala/Prog.scala]: don't make unnecessary whitespace changes

**Comment by garvcho** [comment]
Should I make the changes and do another pull request?

File: generator/src/main/scala/Prog.scala [hunk 4]
```diff
@@ -343,7 +344,24 @@ class Prog(memsize: Int, veccfg: Map[String,String], run_twice: Boolean)
...
+	  val branch_filter = (x: Operand) =>
+      	    x.isInstanceOf[Label] && x.asInstanceOf[Label].label.indexOf("branch_patch") != -1
+    	  val branch_patch = inst.operands.indexWhere(branch_filter)
+	  val jalr_filter1 = (x: Operand) =>
+      	    x.isInstanceOf[Label] && x.asInstanceOf[Label].label.indexOf("jalr_patch1") != -1
+    	  val jalr_patch1 = inst.operands.indexWhere(jalr_filter1)
+	  val jalr_filter2 = (x: Operand) =>
+      	    x.isInstanceOf[Label] && x.asInstanceOf[Label].label.indexOf("jalr_patch2") != -1
+    	  val jalr_patch2 = inst.operands.indexWhere(jalr_filter2)
+    	  if (jalr_patch1 == -1 && branch_patch == -1 && jalr_patch2 == -1){
```
Comment (colinschmidt) [inline generator/src/main/scala/Prog.scala:358]: What is going on here? What are these patch labels?

I also really don't like how you did the filtering, there has to be a better way to do that.

File: generator/src/main/scala/SeqVec.scala [hunk 5]
```diff
@@ -149,6 +149,7 @@ class SeqVec(xregs: HWRegPool, vvregs: HWRegPool, vpregs: HWRegPool, vsregs: HWR
   insts += LUI(xreg_helper, Label("%hi("+vf_init_pred_block.name+")"))
   insts += VF(RegStrImm(xreg_helper, "%lo("+vf_init_pred_block.name+")"))
   
+
```
Comment (colinschmidt) [inline generator/src/main/scala/SeqVec.scala]: again no extra whitespace changes

File: generator/src/main/scala/Prog.scala (hunk 4 above)
Comment (garvcho) [inline generator/src/main/scala/Prog.scala:358]: The filters were there to get rid of the all the jumps in the original segmented code. The original code looked for these "text patches" and filled them in with the correct addresses. I was copying what was done in the original code for the filtering to get rid of those patches and don't know scala well enough to think of a better way.

**Comment by ccelio** [comment]
You can continue to commit to your `add_loops` branch and it will automatically update the pull request. Github is awesome like that.

I sympathize with the desire to run a torture program multiple times (as I've already added the ability with the hacky "run_twice" construct to handle warming the caches), but I'm not quite sold yet on the changes involved in this pull request. 

My first concern is how much of torture's code is being touched by this PR. I'd really rather not remove an entire register from the HWPool - instead, store it out to memory somewhere like how `run_twice` accomplishes this. You might miss bugs that only manifest when using x31 (like Rocket's clever inversion ~rsrc logic to cause x0 to not get synthesized). 

I'm also concerned about the filter code: what does it mean to "remove the jumps"? Is torture no longer testing jumps if "loops" is enabled?

If this code is for you to benchmark processors in primetime, then this may be best served by living as its own fork/branch.

**Comment by garvcho** [comment]
Thanks, I'll try to fix most if not all of these issues. I'll try to implement the suggestion for storing the loop variable in memory as well; I think its a better solution than what's currently being implemented.

**Comment by ben-k** [comment]
This was updated, can someone take a look at it?

@colinschmidt @ccelio

File: generator/src/main/scala/HWRegPool.scala [hunk 6]
```diff
@@ -84,7 +84,7 @@ trait PoolsMaster extends HWRegPool
   }
 }
 
-class XRegsPool extends ScalarRegPool
+class XRegsPool()  extends ScalarRegPool
```
Comment (colinschmidt) [inline generator/src/main/scala/HWRegPool.scala]: why did you need to add empty parens here?

File: generator/src/main/scala/Prog.scala [hunk 7]
```diff
@@ -451,13 +479,15 @@ class Prog(memsize: Int, veccfg: Map[String,String], run_twice: Boolean)
...
-  def output_mem_data() =
+  def output_mem_data(loop_size: Int) =
   {
     var s = "// Memory Blocks\n"
     s += MemDump(core_memory)
     s += "\n"
     s += ".align 8\n"
+    s += "loop_count: .word 0x" + Integer.toHexString(loop_size) + "\n\n"
     s += "execution_count: .word 0x0000000000000001\n\n"
+    //s += "execution_count: .word 0x" + Integer.toHexString(10) + "\n\n"
```
Comment (colinschmidt) [inline generator/src/main/scala/Prog.scala]: Please don't commit commented code.

File: generator/src/main/scala/SeqVec.scala [hunk 8]
```diff
@@ -148,7 +148,7 @@ class SeqVec(xregs: HWRegPool, vvregs: HWRegPool, vpregs: HWRegPool, vsregs: HWR
   extra_code += ProgSegDump(vf_init_pred_block)
   insts += LUI(xreg_helper, Label("%hi("+vf_init_pred_block.name+")"))
   insts += VF(RegStrImm(xreg_helper, "%lo("+vf_init_pred_block.name+")"))
-  
+
```
Comment (colinschmidt) [inline generator/src/main/scala/SeqVec.scala]: Don't make extraneous whitespace changes

**Comment by colinschmidt** [comment]
As far as I can tell run_twice is now obsoleted by this change? If that is indeed the case please remove it.

There are also a few style things I pointed out that should be fixed. I'm happier with the memory storage for the loop variable and if you can fix the other issues would be ok merging this to essentially make run_twice more general.

**Comment by ccelio** [comment]
Yah, generalizing and getting rid of run_twice is a great improvement; for stuff like branch predictors, it may take them a while to "learn" particular patterns.

**Comment by garvcho** [comment]
Its not exact the same as getting rid of run_twice. I'm not entirely sure if this is a big issue but the run_twice option branches back further so that it does the register initializing again. The changes I made jump back to the beginning of the program segment which is after the register initializing. I'm not sure if there is a big difference but I just wanted to point out that there is a slight difference.

**Comment by ccelio** [comment]
> The changes I made jump back to the beginning of the program segment which is after the register initializing.

So long as the register initialization isn't required to make the program perform correctly (i.e., cause a memory access exception), then I don't think this is an issue, and we can remove run_twice.

**Comment by garvcho** [comment]
In that case should I remove it entirely?

**Comment by ccelio** [comment]
Yah, remove run_twice. I assume you've not run into any issues with torture throwing exceptions or crashing out by letting the previous loop iteration remain in the register file when we start the next iteration. I'm fairly certain anything that can go bad will get clobbered anyways, so each iteration will have the same output.

**Comment by colinschmidt** [comment]
Looks good to me now. Thanks for the contribution.

**Comment by garvcho** [comment]
I'm very sorry. I forgot to mention that one of the config names is misleading. There is a config called segment that determines whether to split the program into multiple segments which was the case before I made any chases. The config allows switching between this and just combining on all of the segments into one giant one to avoid empty cycles of jumping around different program segments. If the config is set to true, it combines the programs segments into one and if it is set to false, it leaves them as separate program segments. I realize the name segment might be misleading since setting it false actually breaks it into segments. Should I do something about this?
